### PR TITLE
Changes to work with GHC 8.2.1-rc1 API

### PR DESCRIPTION
**Pull request by AaronFriel**
Todo:

- [x] Compiles on GHC 8.2.1-rc2 
- [x] Compiles on GHC 8.0.2
- [x] Compiles on GHC 7.10.3

### CODE REVIEW

**Comment by AaronFriel** [comment]
@purcell Uhm, hi, I'm just going to tag you because you recently committed, and I don't know the contributor standards/contributor license agreement stuff on this project. I'd like to know what (if anything) I should do at this point. This is a very "Works On My Machine" patch, and I'm not at all familiar with the Intero internals. I know that I don't know enough to make this a much better patch.

**Comment by purcell** [comment]
I've been sticking mostly to the elisp side of things, and I guess one of the haskell-side committers would need to pick this up and review/sponsor the change, sorry.

File: cbits/hschooks.c [hunk 1]
```diff
@@ -34,7 +34,7 @@ defaultsHook (void)
     // This helps particularly with large compiles, but didn't work
     // very well with earlier GHCs because it caused large amounts of
     // fragmentation.  See rts/sm/BlockAlloc.c:allocLargeChunk().
-    RtsFlags.GcFlags.heapSizeSuggestionAuto = rtsTrue;
```
Comment (chrisdone) [inline cbits/hschooks.c]: I think this ought to have a CPP flag for backwards-compat.

File: src/GhciFind.hs [hunk 2]
```diff
@@ -50,7 +50,7 @@ findQualifiedSource importDecls sample =
             Just name
           | otherwise = Nothing
           where name = unLoc (ideclName m)
-                asName = fmap moduleNameString . ideclAs
+                asName = fmap (moduleNameString . unLoc) . ideclAs
```
Comment (chrisdone) [inline src/GhciFind.hs]: This should have a CPP if.

File: src/GhciFind.hs [hunk 3]
```diff
@@ -364,7 +364,7 @@ findType infos fp string sl sc el ec =
                             Nothing -> return (FindType minfo ty)
                   _ ->
                     fmap (FindType minfo)
-                         (exprType string)
+                         (exprType TM_Inst string)
```
Comment (chrisdone) [inline src/GhciFind.hs]: Again, needs CPP.

File: src/GhciMonad.hs [hunk 4]
```diff
@@ -321,14 +321,14 @@ runDecls decls = do
...
-resume :: (SrcSpan -> Bool) -> GHC.SingleStep -> GHCi GHC.RunResult
+resume :: (SrcSpan -> Bool) -> GHC.SingleStep -> GHCi GHC.ExecResult
 resume canLogSpan step = do
   st <- getGHCiState
   reifyGHCi $ \x ->
     withProgName (progname st) $
     withArgs (args st) $
       reflectGHCi x $ do
-        GHC.resume canLogSpan step
+        GHC.resumeExec canLogSpan step
```
Comment (chrisdone) [inline src/GhciMonad.hs]: This whole function should also have some CPP.

**Comment by AaronFriel** [comment]
@chrisdone:
> The way to do this is to write e.g. ...
> 
> ```
> #if __GLASGOW_HASKELL__ >= 802
> new code
> #else
> old code
> #endif
> ```

This would need to be `821`, correct, due to the changes in the GHC API? Anyhow, it did not work for me. Perhaps it was just the version of GHC 8.2.1-rc1 I was using? I tried using ">= 821" on another project.

File: cbits/hschooks.c (hunk 1 above)
Comment (AaronFriel) [inline cbits/hschooks.c]: Is this flag no longer used in 8.2.1? I could find almost no documentation on it.

**Comment by chrisdone** [comment]
@AaronFriel I believe the GHC number 802 refers to 8.2 and that you cannot test for the minor version because it _should_ not matter. See here:

https://downloads.haskell.org/~ghc/7.0.3/docs/html/users_guide/version-numbering.html

> The value of __GLASGOW_HASKELL__ (see Section 4.11.3, “Options affecting the C pre-processor”) for a major release x.y.z is the integer xyy (if y is a single digit, then a leading zero is added, so for example in version 6.8.2 of GHC we would have __GLASGOW_HASKELL__==608).

**Comment by vagarenko** [comment]
Can't compile it with GHC-8.2.0rc2
```
src\InteractiveUI.hs:3461:37: error:
    * Variable not in scope:
        utf8DecodeString :: Ptr b0 -> Int -> IO String
    * Perhaps you meant one of these:
        `utf8EncodeString' (imported from Encoding),
        `utf8DecodeByteString' (imported from Encoding),
        `utf8DecodeStringLazy' (imported from Encoding)
     |
3461 |                         $ \(p,n) -> utf8DecodeString (castPtr p) n
     |                                     ^^^^^^^^^^^^^^^^
```
Seems like `utf8DecodeString` was removed from GHC API.

**Comment by vagarenko** [comment]
Yes, it was removed here:
https://github.com/ghc/ghc/commit/1cc82d38759c7a5f527ccc6cb514b8ba576cc3d1

**Comment by AaronFriel** [comment]
@chrisdone Oh well that makes sense!

**Comment by AaronFriel** [comment]
I don't understand how you have "-Werror" in that test script, I can't run it without it erroring on minor warnings in dependencies, e.g.: tabs in `random`.

**Comment by chrisdone** [comment]
I'm going to try adding 8.2.1 to travis.

**Comment by chrisdone** [comment]
Great work @AaronFriel. People will be very happy when 8.2.1 lands. 👏 😄

**Comment by purcell** [comment]
Yes, thanks Aaron!

**Comment by AaronFriel** [comment]
Thank you!

**Pull request merged**